docs/source/notes/rref.rst

```diff
@@ -0,0 +1,304 @@
+instances, i.e.,
+when ``OwnerRRef`` is deleted, all ``UserRRef`` instances could be either indeed
+deleted or unknown. The dangerous case is when some forks are unknown and others
+are deleted.
+
+**G2** trivially guarantees that no parent ``UserRRef`` can be deleted before
+the owner knows all of its children ``UserRRef`` instances. However, it is
+possible that the child ``UserRRef`` may be deleted before the owner knows its
+parent ``UserRRef``.
+
+Consider the following example, where the ``OwnerRRef`` forks to A, then A forks
+to Y, and Y forks to Z.:
+
+.. code::
+
+  OwnerRRef -> A -> Y -> Z
+
+If all of Z's messages, including the delete message, are processed by the
+owner before all messages from Y, the owner will learn Z's deletion before
+knowing Y. Nevertheless, this does not cause any problem. Because, at least
+one of Y's ancestors will be alive (in this case, A) and it will
+prevent the owner from deleting the ``OwnerRRef``. More specifically, if the
+owner does not know Y, A cannot be deleted due to **G2**, and the owner knows A
+as the owner is A's parent.
+
+Things get a little trickier if the RRef is created on a user:
+
+
+.. code::
+
+  OwnerRRef
+      ^
+      |
+      A -> Y -> Z
+
+
+If Z calls to_here on the ``UserRRef``, the owner at least knows A when Z is
+deleted, because otherwise, ``to_here`` wouldn't finish. If Z does not call
+to_here, it is possible that the owner receives all messages from Z before
+any message from A and Y. In this case, as the real data of the ``OwnerRRef``
+has not been created yet, there is nothing to be deleted either. It is the same
+as Z does not exist at all. Hence, it's still OK.
+
+Implementation
+--------------
+
+**G1** is implemented by sending out a delete message in ``UserRRef``
+destructor. To provide **G2**, the parent ``UserRRef`` is put into a context
+whenever it is forked, indexed by the new ``ForkId``. The parent ``UserRRef`` is
+only removed from the context when it receives an ACK from the child, and the
+child will only send out the ACK when it is confirmed by the owner.
+
+
+Protocol Scenarios
+^^^^^^^^^^^^^^^^^^
+
+Let's now discuss how the above designs translate to the protocol in four
+scenarios.
+
+User Share RRef with Owner as Return Value
+------------------------------------------
+
+
+.. code::
+  import torch
+  import torch.distributed.rpc as rpc
+
+  # on worker A
+  rref = rpc.remote('B', torch.add, args=(torch.ones(2), 1))
+  # say the rref has RRefId 100 and ForkId 1
+  rref.to_here()
+
+
+In this case, the ``UserRRef`` is created on the user worker A, then it is
+passed to the owner worker B together with the remote message, and then B
+creates the ``OwnerRRef``. The method ``rpc.remote`` returns immediately,
+meaning that the ``UserRRef`` can be forked/used before the owner knows about
+it.
+
+On the owner, when receiving the ``rpc.remote`` call, it will create the
+``OwnerRRef``, and returns an ACK to acknowledge ``{100, 1}``. Only
```

Comment:
Why the values {100,1}? Is it RRefId 100 and ForkId 1?

Reply:
Yes, let me add that